### PR TITLE
Cleanup: replace go-ethereum ecies for umbracle/ecies

### PR DESCRIPTION
**Pull request by vkhoroz**
A chain of commits which gradually phases out our go-ethereum dependency.
The rationale for each change is explained in the commit messages.
The tests have passes on each step, warranting the inter-operability with old implementation.

There is still a huge copy-paste-amend function hanging out in the ecies_pkcs11.go module.
Ideally, I think we should fork the umbracle/ecies project (which is hardly active BTW), inject our pkcs11 code into that fork and maintain it.
That would allow to re-use 98% of the common ECIES code in fioctl/device-gateway via import rather than copy-paste.

A potentially viable path would be to merge this, fork-amend umbracle/ecies, and then reuse it here as the next commit.

Signed-off-by: Volodymyr Khoroz <volodymyr.khoroz@foundries.io>

### CODE REVIEW

**Comment by vkhoroz** [comment]
@doanac @mike-sul I have uploaded quite some PRs on ota-lite & device-gateway. So on my Thursday evening and Friday before the New Year I had a dilemma what to do next.

This phasing out of go-ethereum seemed like enough standalone task with clear value.
It should give peace of mind to the device owners.

File: internal/ecies.go [hunk 1]
```diff
@@ -36,8 +36,7 @@ func (ec *EciesCrypto) Decrypt(value string) ([]byte, error) {
 }
 
 func (ec *EciesCrypto) Encrypt(value string) (string, error) {
-	pub := ecies.ImportECDSAPublic(ec.PrivKey.Public())
-	enc, err := ecies.Encrypt(rand.Reader, pub, []byte(value), nil, nil)
+	enc, err := ecies.Encrypt(rand.Reader, ec.PrivKey.Public(), []byte(value), nil, nil)
```
Comment (StealthyCoder) [inline internal/ecies.go:39]: Nice that we can immediately pass along the `Public` rather than doing a round about step.

File: internal/ecies_pkcs11.go [hunk 2]
```diff
@@ -216,38 +146,44 @@ func EciesDecrypt(prv PrivateKey, c, s1, s2 []byte) (m []byte, err error) {
...
+		return nil, ecies.ErrInvalidMessage
 	}
 	if subtle.ConstantTimeCompare(c[mEnd:], d) != 1 {
-		return nil, ErrInvalidMessage
+		return nil, ecies.ErrInvalidMessage
 	}
 
-	return symDecrypt(params, Ke, c[mStart:mEnd])
+	// rand is unused below, but mandatory per some legacy interface
+	return symDecrypt(nil, params, Ke, c[mStart:mEnd])
```
Comment (StealthyCoder) [inline internal/ecies_pkcs11.go:188]: It is not even used by the implementation in ecies lib.

**Review by StealthyCoder — APPROVED** [review]

**Comment by StealthyCoder** [comment]
> There is still a huge copy-paste-amend function hanging out in the ecies_pkcs11.go module. Ideally, I think we should fork the umbracle/ecies project (which is hardly active BTW), inject our pkcs11 code into that fork and maintain it. That would allow to re-use 98% of the common ECIES code in fioctl/device-gateway via import rather than copy-paste.
> 
> A potentially viable path would be to merge this, fork-amend umbracle/ecies, and then reuse it here as the next commit.

I would like this very much, we can even do small quality of life things like removing `rand`  from `symDecrypt`. Makes everything more streamlined. :muscle:

File: internal/ecies_params_ethereum.go [hunk 3]
```diff
@@ -107,29 +86,15 @@ var (
 )
 
 var paramsFromCurve = map[elliptic.Curve]*ECIESParams{
-	ethcrypto.S256(): ECIES_AES128_SHA256,
```
Comment (mike-sul) [inline internal/ecies_params_ethereum.go:110]: Potentially, "The Ethereum's S256 (which corresponds to secp256k1)" could have been used by some customer that generated their device certificate "offline" (and signed with offline CA) without `lmp-dev-reg`. I think is is very unlikely though.

File: go.mod [hunk 4]
```diff
@@ -5,29 +5,25 @@ go 1.17
 require (
 	github.com/ThalesIgnite/crypto11 v1.2.5
 	github.com/coreos/go-systemd/v22 v22.5.0
-	github.com/ethereum/go-ethereum v1.10.22
 	github.com/google/uuid v1.2.0
 	github.com/miekg/pkcs11 v1.0.3-0.20190429190417-a667d056470f
 	github.com/pelletier/go-toml v1.8.0
 	github.com/stretchr/testify v1.7.2
+	github.com/umbracle/ecies v0.0.0-20190516151710-4bface869a50
```
Comment (mike-sul) [inline go.mod:12]: It looks like the `umbracle/ecies` is fork from `https://github.com/obscuren/ecies`, why not use the original one?

Comment (mike-sul) [reply go.mod:12]: The license is BSD-3.

Comment (StealthyCoder) [reply go.mod:12]: > This implementation is direct fork of Kylom's implementation. I claim no authorship over this code apart from some minor modifications. Please be aware this code **has not yet been reviewed**. 

So do we know if the other code has been reviewed? 😊

File: internal/ecies_params_ethereum.go (hunk 3 above)
Comment (StealthyCoder) [reply internal/ecies_params_ethereum.go:110]: If we wanted to we could use: https://github.com/decred/dcrd/tree/master/dcrec/secp256k1 to bring it back in without relying on go-ethereum. It has the copyfree ISC license.

File: internal/ecies_ethereum.go [hunk 5]
```diff
@@ -91,30 +91,9 @@ func (prv *PrivateKeyPkcs11) Public() *ecies.PublicKey {
...
-	return k[:kdLen]
-}
-
-// roundup rounds size up to the next multiple of blocksize.
-func roundup(size, blocksize int) int {
-	return size + blocksize - (size % blocksize)
-}
+//
+//go:linkname concatKDF github.com/umbracle/ecies.concatKDF
+func concatKDF(hash hash.Hash, z, s1 []byte, kdLen int) []byte
```
Comment (mike-sul) [inline internal/ecies_ethereum.go:96]: looks like a hack :).

File: .jobserv.yml [hunk 6]
```diff
@@ -4,7 +4,7 @@ triggers:
     type: github_pr
     runs:
       - name: unit-test
-        container: golangci/golangci-lint:v1.48
```
Comment (mike-sul) [inline .jobserv.yml:7]: Probably this commit should be extracted into the separate PR?

File: go.mod [hunk 7]
```diff
@@ -1,6 +1,6 @@
 module github.com/foundriesio/fioconfig
 
-go 1.17
```
Comment (mike-sul) [inline go.mod:3]: Probably this commit should be extracted into the separate PR?

File: Makefile [hunk 8]
```diff
@@ -24,7 +24,7 @@ format:
 	@gofmt -l  -w ./
 
 check: test
-	@test -z $(shell gofmt -l ./ | tee /dev/stderr) || (echo "[WARN] Fix formatting issues with 'make fmt'"; exit 1)
```
Comment (mike-sul) [inline Makefile:27]: Probably this commit should be extracted into the separate PR?

**Review by mike-sul — COMMENTED** [review]
I think the last three commits should be extracted into a separate PR.

**Comment by mike-sul** [comment]
I think the key question is: "Can we trust `umbracle/ecies`"?
If so, then we should change `device-gateway` and fioctl accordingly too.

**Comment by StealthyCoder** [comment]
> I think the key question is: "Can we trust `umbracle/ecies`"? If so, then we should change `device-gateway` and fioctl accordingly too.

I looked more into the `go-ethereum` ecies code and it uses the exact same code except for the sec256k1 part. So if we can trust the `go-ethereum` code I see no reason why we cannot trust `umbracle/ecies`. :slightly_smiling_face:

File: internal/ecies_pkcs11.go (hunk 2 above)
Comment (vkhoroz) [reply internal/ecies_pkcs11.go:188]: yep... it looks like some rudiment; so much was my surprise when I saw that argument.

File: internal/ecies_params_ethereum.go (hunk 3 above)
Comment (vkhoroz) [reply internal/ecies_params_ethereum.go:110]: Unfortunately, the decred above uses the go-ethereum under the hood, as do all other 3PP libraries implementing this algorithm except the https://github.com/ThePiachu/Golang-Koblitz-elliptic-curve-DSA-library/blob/master/bitelliptic/bitelliptic.go#L326. But that was not maintained for the last 8 years and is not even compatible with the Golang's crypto interfaces.

So, initially I wrote the code to support the S256 by copying the curve numbers from the spec and using reflection to create internal Golang crypto objects. And it worked well (I could encrypt/decrypt data in go-ethereum compatible way. I still have that code around, but it is enough of a contemporary programmer's nightmare to not share it anywhere.

But, the truth is: no single customer could ever use the S256 curve, as it is not supported by the Golang's TLS implementation. Such certificates would simply fail as "not supported EC algorithm" when reading such client certificates.

I thuought I was enough explicit about that in the commit message. Please, let me know if you read it and it is not clear enough.

File: go.mod (hunk 4 above)
Comment (vkhoroz) [reply go.mod:12]: There is a dozen of forks on the `obscuren/ecies`, which itself is a fork from `go-ethereum`, which is a work by Google engineer granted to `go-ethereum` under the Golang-compatible Google's license.

But only the `umbracle/ecies` seemed like maintaned at least 3 years ago and have some fixes which make it work with the recent Golang versions. All other forks, as well, as the original are practically dead ends for many years.

It looks like everybody [in need of ECIES and using Golang] cares not about the license or the import of universe. So all well maintained OS projects I've seen so far simply use the `go-ethereum`'s implementation of ECIES.

File: internal/ecies_ethereum.go (hunk 5 above)
Comment (vkhoroz) [reply internal/ecies_ethereum.go:96]: Yes, it is... still the Golang standard library uses it extensively. So we should treat it as safe from any perspective.
Every time I tried to do something wrong on this line - either the compiler blamed on me, or the code panicked. Both ways to fail seem good to me.

Comment (vkhoroz) [reply internal/ecies_ethereum.go:96]: Anyway, I consider this temporary if we go for forking `umbracle/ecies` and injecting our pkcs11 changes into it.

File: .jobserv.yml (hunk 6 above)
Comment (vkhoroz) [reply .jobserv.yml:7]: Probably, you're right.

The CI was failing for me about bad format (while locally the auto-formatter worked well), so I was guessing about the difference between the formatter/linter/golang versions here and there.

Now that I made the `make check` pass in the CI, I could try to extract these non-related commits into a separate PR.

File: go.mod (hunk 7 above)
Comment (vkhoroz) [reply go.mod:3]: ok

File: Makefile (hunk 8 above)
Comment (vkhoroz) [reply Makefile:27]: ok

**Comment by vkhoroz** [comment]
> > I think the key question is: "Can we trust `umbracle/ecies`"? If so, then we should change `device-gateway` and fioctl accordingly too.
> 
> I looked more into the `go-ethereum` ecies code and it uses the exact same code except for the sec256k1 part. So if we can trust the `go-ethereum` code I see no reason why we cannot trust `umbracle/ecies`. 🙂

There is one particular reason why I don't trust umbracle/ecies: its repo uses no tags, so the `go mod tidy` literally pointed at a random commit in master. Once we fork it - that little nuissance will be gone. From all other perspectives it is almost a clone of the `go-ethereum`'s ECIES implementation... made to work with the standard Golang crypto library.

File: internal/ecies_params_ethereum.go (hunk 3 above)
Comment (mike-sul) [reply internal/ecies_params_ethereum.go:110]: In the case of an offline device certificate we need to make sure that both ota-lite and device-gateway won't accept it.

Comment (doanac) [reply internal/ecies_params_ethereum.go:110]: i think i just copied this file as close to "as-is" as possible at the time. I wasn't sure how much it might change over time and just wanted to make sure it would be easy to pull in changes. Turns out there have never been changes.

File: go.mod [hunk 9]
```diff
@@ -5,29 +5,25 @@ go 1.17
 require (
```
Comment (doanac) [inline go.mod:5]: can't comment on a commit message but this change is really nice. i think you real genius here was figuring out we didn't need need ethereum curve and that was the real baggage we'd been caring around unnecessarily?

File: internal/ecies_params_ethereum.go (hunk 3 above)
Comment (vkhoroz) [reply internal/ecies_params_ethereum.go:110]: > In the case of an offline device certificate we need to make sure that both ota-lite and device-gateway won't accept it.

I don't think so.

There are 100500 ways how a user could craft a particular certificate in a way to make it pass the ota-lite validation but fail the TLS connection. But, as long as the device-gateway will not accept a TLS connection with some client certificate - that certificate is deemed unsupported from our perspective.

By generating the device certificate using a custom tool (not vanilla lmp-device-register), the customer steps into a danger zone; so they are responsible to verify that their solution works end-2-end.

**Review by doanac — APPROVED** [review]
this is cool. the commit with the linking hacks is a little wild, but I think i can stomach it because we can easily prove this works.

File: go.mod (hunk 9 above)
Comment (vkhoroz) [reply go.mod:5]: yep, that every word is true.

TBH I've spent a half day to re-implement that curve (repeating myself)... just to find out that it fails on Golang's TLS implementation... so, the further search led me to understand that we use the P256 curve which has a misleadingly close name to S256.

Unfortunately for us, the `go-ethereum` crypto library is as cryptic as it could be when naming curves - one should look at the curve numbers and compare them to industry standards in order to guess, rather than just read their name.

File: internal/ecies_params_ethereum.go (hunk 3 above)
Comment (StealthyCoder) [reply internal/ecies_params_ethereum.go:110]: > I thought I was explicit about that in the commit message. Please, let me know if you read it and it is not clear enough.

Actually my apologies, I forgot to read that first. That message is very clear and perfect.
